internal/confirmations/confirmed_block_listener.go

```diff
@@ -0,0 +1,317 @@
+		log.L(cbl.ctx).Debugf("Notification of block %d/%s < fromBlock %d", block.BlockNumber, block.BlockHash, cbl.fromBlock)
+		return
+	}
+
+	// If the block is before our checkpoint, we ignore it completely
+	if cbl.rollingCheckpoint != nil && block.BlockNumber.Uint64() <= cbl.rollingCheckpoint.Block {
+		log.L(cbl.ctx).Debugf("Notification of block %d/%s <= checkpoint %d", block.BlockNumber, block.BlockHash, cbl.rollingCheckpoint.Block)
+		return
+	}
+
+	// Otherwise see if it's a conflicting fork to any of our existing blocks
+	for idx, existingBlock := range cbl.blocksSinceCheckpoint {
+		if existingBlock.BlockNumber == block.BlockNumber {
+			// Must discard up to this point
+			cbl.blocksSinceCheckpoint = cbl.blocksSinceCheckpoint[0:idx]
+			// This block fits, and add this on the end.
+			if idx == 0 || block.ParentHash == cbl.blocksSinceCheckpoint[idx-1].BlockHash {
+				log.L(cbl.ctx).Debugf("Notification of block %d/%s after block %d/%s", block.BlockNumber, block.BlockHash, existingBlock.BlockNumber, existingBlock.BlockHash)
+				cbl.blocksSinceCheckpoint = append(cbl.blocksSinceCheckpoint[0:idx], block)
+			} else {
+				log.L(cbl.ctx).Debugf("Notification of block %d/%s conflicting with previous block %d/%s", block.BlockNumber, block.BlockHash, existingBlock.BlockNumber, existingBlock.BlockHash)
+			}
+			break
+		}
+	}
+
+	// We didn't fit the block into our existing tree, so this means it's ahead of where we are up to.
+	// So just ensure the dispatcher is racing up to it
+	cbl.tapDispatcher()
+
+}
+
+func (cbl *confirmedBlockListener) tapDispatcher() {
+	select {
+	case cbl.dispatcherTap <- struct{}{}:
+	default:
+	}
+}
+
+func (cbl *confirmedBlockListener) dispatcher() {
+	defer close(cbl.dispatcherDone)
+
+	for {
+		if !cbl.waitingForFromBlock {
+			// tight spin getting blocks until we it looks like we need to wait for a notification
+			for cbl.getNextBlock() {
+				// In all cases we ensure that we move our confirmation window forwards.
+				// The checkpoint block is always final, and we never move backwards
+				cbl.dispatchAllConfirmed()
+			}
+		}
+
+		select {
+		case <-cbl.dispatcherTap:
+		case <-cbl.ctx.Done():
+			log.L(cbl.ctx).Debugf("Confirmed block dispatcher stopping")
+			return
+		}
+
+	}
+}
+
+func (cbl *confirmedBlockListener) getNextBlock() (more bool) {
+
+	var nextBlock *apitypes.BlockInfo
+	err := cbl.retry.Do(cbl.ctx, "next block", func(_ int) (retry bool, err error) {
+		// Find the highest block in the lock
+		cbl.stateLock.Lock()
+		blockNumberToFetch := cbl.fromBlock
+		if cbl.rollingCheckpoint != nil && cbl.rollingCheckpoint.Block >= cbl.fromBlock {
+			blockNumberToFetch = cbl.rollingCheckpoint.Block + 1
+		}
+		if len(cbl.blocksSinceCheckpoint) > 0 {
+			blockNumberToFetch = cbl.blocksSinceCheckpoint[len(cbl.blocksSinceCheckpoint)-1].BlockNumber.Uint64() + 1
+		}
+		cbl.stateLock.Unlock()
+
+		// Get the next block
+		nextBlock, err = cbl.bcm.getBlockByNumber(blockNumberToFetch, false, "")
```

Comment:
@peterbroadhurst I feel this could be too aggressive to avoid different block info with the same Block Number. The number of requests TM sends to a JSON-RPC endpoint affects performance a lot so we should keep it efficient.

The logic need to balance between the following choices when there are multiple block listeners in flight:
1. listeners should go and revalidate the information of each block from the "fromBlock" even if they are validating the same block at the same time.
2. listeners should use cached information of blocks when possible.

I can see the second one won't work with more logic added as we'll stuck at unmatching block parent hash when the cached block is outdated.

However, I do also feel the first one is a bit too aggressive if each listener will need to fetch each block with no cache at all. For example, if we have 10 listener, then we are emitting 10 get blockInfo call to the JSON-RPC endpoint for every single block.

Reply:
One solution could be default to use cache, and stop using cache when we detected the parent hash do not match the one in the unconfirmed previous block.

Reply:
I agree with all the desires and concerns of things to think about here @Chengxuan - so thank you for raising. 

I completely agree it would be great _if_ we could work out a way to have a cache on this line, but I actually I found I had to _disable the previously existing cache_ here because there's no way I could pre-empt knowing what re-org has happened on the node.

However, please note this **only kicks in** if the block addition by hash (which is cached) doesn't work. If we're just adding hashed block to the list as they arrive, none of the listeners will query at all. That's happening on the other go routine.

The optimization for block arriving in order from the chain to multiple listeners happens here (and is cache-enabled, because it's a get-by-hash which is deterministic to cache):
https://github.com/hyperledger/firefly-transaction-manager/blob/072aa386c57c784f02850948dd10dc0c6c04ff00/internal/confirmations/confirmed_block_listener.go#L196-L198

Reply:
Oh, just a minute. That comment of `// This block fits, and add this on the end.` reflects what I described above, but the actual code in the notification is only dealing with the re-org case.

Sorry - I must have forgotten to come back to that code to put in the add-to-end case in after I did some work.
I'll complete that, and then the code will work as I describe in my comment here.

You have also made me think of another optimization - the dispatcher (or "catch-up" routine) is always asking for blocks until it gets an error. This means in the caught-up notification cases, it's always doing an inefficient extra JSON/RPC call just to find out there's no new blocks.

If I maintain a "last notification height" on the notification listener (or similar), then we can optimize such that if the dispatch is caught up with the head of the chain, then there's no immediate re-poll necessary.

I'll do both of these and request a re-review.

Reply:
Ok - that's was harder than I'd thought! Logic got significantly more complex, to deal with the locking and avoid a situation where any go channel gets unnecessarily blocked.

However, I believe I now have it so that in the case the listener is keeping up, there will only be cached `eth_getBlockByHash` calls, and no `eth_getBlockByNumber` calls.